patches/tModLoader/Terraria/Main.cs.patch

```diff
@@ -3239,13 +3239,44 @@
  					}
  				}
  			}
-@@ -36640,11 +_,13 @@
- 							if (Main.mouseLeftRelease && Main.mouseLeft && Main.player[Main.myPlayer].BuyItem(num62, -1))
+@@ -36521,12 +_,18 @@
+ 					string text2 = Lang.inter[46].Value + ": ";
+ 					if (Main.reforgeItem.type > 0)
+ 					{
++						bool canApplyDiscount = true;
+ 						int num62 = Main.reforgeItem.value;
+-						if (Main.player[Main.myPlayer].discount)
+-						{
+-							num62 = (int)((double)num62 * 0.8);
+-						}
+-						num62 /= 3;
++
++						if (ItemLoader.ReforgePrice(Main.reforgeItem, ref num62, ref canApplyDiscount))
++						{
++							if (canApplyDiscount && Main.player[Main.myPlayer].discount)
++							{
++								num62 = (int)((double)num62 * 0.8);
++							}
++							num62 /= 3;
++						}
++
+ 						string text3 = "";
+ 						int num63 = 0;
+ 						int num64 = 0;
+@@ -36637,14 +_,17 @@
+ 							}
+ 							Main.mouseReforge = true;
+ 							Main.player[Main.myPlayer].mouseInterface = true;
+-							if (Main.mouseLeftRelease && Main.mouseLeft && Main.player[Main.myPlayer].BuyItem(num62, -1))
++							if (Main.mouseLeftRelease && Main.mouseLeft && Main.player[Main.myPlayer].BuyItem(num62, -1) && ItemLoader.PreReforge(Main.reforgeItem))
```

Comment:
Woops. `Main.player[Main.myPlayer].BuyItem` will take away gold regardless if the PreReforge hook returns true or not.